File: Insertion_Sort/Insertion_Sort.kt
//Function for sorting an list using Insertion Sort
fun Insertion_sort(list: MutableList<Int>): List<Int> {
    for(i in 0 until list.size) {
        var key = list[i]
        var j = i - 1
        /*Move elements of arr[0..i-1], that are
          greater than key, to one position ahead
          of their current position*/
        while(j > 0 && list[j] > key) {
            list[j+1] = list[j]
            j = j - 1
        }
        list[j+1] = key
    }
    return list
}

//Driver function for Insertion sort
fun main(args: Array<String>) {
    val list = mutableListOf(4, 67, 89, 2, 45, 102)
    println("Unsorted List: $list")
    val sortedlist = Insertion_sort(list)
    println("Sorted List: $sortedlist")
}

//Input: Unsorted: [4, 67, 89, 2, 45, 102]
//Output: Sorted: [2, 4, 25, 67, 89, 102]

File: Algo_Ds_Notes/Insertion_Sort/Insertion_Sort.c
#include <stdio.h>

// Function for insertion sort
void Insertion_Sort(int array[], int size)
{
    int temp, j, i;

    for(i = 1; i < size; i++)
    {
        temp = array[i];
        j = i - 1;

        // Do swapping
        while(j >= 0 && array[j] > temp)
        {
            array[j + 1] = array[j];
            j--;
        }

        array[j + 1] = temp;
    }
}

// Function to print elements of array
void Print_Array(int array[], int size)
{
    for(int i = 0; i < size; i++)
        printf("%d\t",array[i]);

    printf("\n");
}

// Driver Function
int main()
{
    int num;
    scanf("%d", &num);
    int array[num];
    for (int i = 0; i < num; i++) {
        scanf("%d", &array[i]);
    }
    Insertion_Sort(array, num);
    Print_Array(array, num);
    return 0;
}

/*
Input :
num = 6
array = [1, 6, 3, 3, 5, 2]

Output :
[1, 2, 3, 3, 5, 6]
*/
